[tt-train]: Standalone build fails with GCC due to missing compiler selection logic

<!--
Bug Report for tt-metal
Title: [tt-train]: Standalone build fails with GCC due to missing compiler selection logic
Labels: bug
-->

## Issue Metadata

### Component / Area

tt-train

### Issue Type (optional)

Other

## Issue Description

### Observed

When building tt-train standalone via `pip install -e tt-train`, the build fails with compiler errors in the `reflect` library dependency. The build uses system default GCC instead of Clang-20, and GCC 11 has a bug with `constexpr` lambda function pointers used in the `reflect` library.

### Expected

Standalone tt-train builds should automatically use Clang-20 (as they did before commit `a7618b9282`), or at minimum validate that the compiler meets requirements (GCC 12+ or Clang 15+).

## Steps to Reproduce the Issue

### 1. Steps (exact commands)

In a containerized build environment (Ubuntu 22.04 base with GCC 11 as system default and Clang-20 available):

```bash
# Clone tt-metal at a commit after a7618b9282
git clone --recursive https://github.com/tenstorrent/tt-metal.git
cd tt-metal
git checkout 152f82cdd7
git submodule update --init --recursive
git lfs pull
git submodule foreach 'git lfs pull'

# Build tt-metal (succeeds - build_metal.sh handles compiler selection)
source python_env/bin/activate
./build_metal.sh --debug --build-all

# Install tt-train via pip (fails - uses system GCC instead of Clang-20)
pip install -e tt-train
```

The issue is that `pip install -e tt-train` triggers a separate CMake configuration for tt-train, which no longer has compiler selection logic and falls back to system default GCC instead of using Clang-20.

### 2. Input data / link or description

No special input data required. Issue occurs during compilation of tt-train when installed as a standalone pip package.

### 3. Frequency

Always (100% reproducible when using GCC < 12 as system default compiler)

## System Details

### 1. Software Versions

- OS version: Ubuntu 22.04 (Jammy)
- GCC version: 11.4.0 (system default via `/usr/bin/g++`)
- Clang version: 20 (available but not auto-selected for tt-train standalone build)
- tt-metal commit: `152f82cdd7` or any commit after `a7618b9282`
- reflect library: v1.2.6 (fetched via CPM)

### 2. Hardware Details

- Product: N/A (build-time issue, no hardware required)
- Card/System: N/A

## Regression Info (Optional)

### Is this a regression?

Yes

### Regression Details

- First bad version: `a7618b9282` ("TT-Train: bump clang version from 17 to 20 #36568")
- Last known good version: `4cca23f9ad` ("Migrate ttml metal ops to free function style #36301")
- Git bisect status: Completed - regression introduced by `a7618b9282`

## Logs & Diagnostics (Optional)

### Logs & Diagnostics

```
/workspace/tt-metal/tt-train/build/_deps/reflect-src/reflect:1318:35: error: non-constant condition for static assertion
 1318 |     static_assert(type_id_v<int>  != type_id_v<void>);
      |                   ~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~
/workspace/tt-metal/tt-train/build/_deps/reflect-src/reflect:1318:35: error: '(<lambda()>::_FUN != <lambda()>::_FUN)' is not a constant expression
/workspace/tt-metal/tt-train/build/_deps/reflect-src/reflect:1320:35: error: non-constant condition for static assertion
 1320 |     static_assert(type_id_v<int>  != type_id_v<const int&>);
      |                   ~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~
/workspace/tt-metal/tt-train/build/_deps/reflect-src/reflect:1320:35: error: '(<lambda()>::_FUN != <lambda()>::_FUN)' is not a constant expression
/workspace/tt-metal/tt-train/build/_deps/reflect-src/reflect:989: confused by earlier errors, bailing out
ninja: build stopped: subcommand failed.

*** CMake build failed
```

The error occurs because:
1. Commit `a7618b9282` removed `tt-train/cmake/compilers.cmake` and the compiler selection logic from `tt-train/CMakeLists.txt`
2. Without this logic, standalone builds use the system default compiler (GCC 11)
3. The `reflect` library (v1.2.6) requires GCC 12+ or Clang 15+ due to `constexpr` lambda function pointer handling
4. GCC 11 has a known bug that causes these `static_assert` failures

## Impact & Priority (Optional)

### Priority

P2

### Impact

- Affected workflows: Any workflow that builds tt-train standalone via `pip install -e tt-train` on systems where GCC < 12 is the default compiler
- Release or date risk: Blocks standalone tt-train development on Ubuntu 22.04 with default packages

--- a/tt-train/CMakeLists.txt
+++ b/tt-train/CMakeLists.txt
@@ -1,4 +1,18 @@
 cmake_minimum_required(VERSION 3.18...4.2)
+include(cmake/compilers.cmake)
+
+# When building as a standalone top-level project, inherit the compiler from
+# tt-metal's CMakeCache.txt for ABI compatibility. This is necessary because
+# build systems like scikit-build-core (used by "pip install -e") set CC/CXX
+# to generic values that don't match tt-metal's compiler.
+#
+# To use a different compiler, disable inheritance:
+#   cmake -DTT_TRAIN_INHERIT_COMPILER=OFF -DCMAKE_C_COMPILER=gcc-12 ...
+option(TT_TRAIN_INHERIT_COMPILER "Inherit compiler from tt-metal's CMakeCache.txt" ON)
+
+if(CMAKE_SOURCE_DIR STREQUAL CMAKE_CURRENT_SOURCE_DIR AND TT_TRAIN_INHERIT_COMPILER)
+    inherit_compiler_from_tt_metal()
+endif()
 
 project(
     ml-framework-cpp
@@ -6,6 +20,7 @@ project(
         C
         CXX
 )
+CHECK_COMPILERS()
 
 message(STATUS "Build type: ${CMAKE_BUILD_TYPE}")
 set(CMAKE_CXX_FLAGS_RELEASE "-O3")

--- a/tt-train/cmake/compilers.cmake
+++ b/tt-train/cmake/compilers.cmake
@@ -1,0 +1,88 @@
+# Compiler selection and validation for tt-train.
+#
+# When building standalone (not as a tt-metal subproject), tt-train needs
+# a compiler compatible with the one used to build tt-metal, since it links
+# against tt-metal libraries (libtt_metal.so, _ttnncpp.so).
+#
+# Compiler selection (standalone top-level builds):
+#   - By default, the compiler is inherited from tt-metal's CMakeCache.txt
+#     to ensure ABI compatibility (TT_TRAIN_INHERIT_COMPILER=ON).
+#   - To use a different compiler, disable inheritance and set explicitly:
+#     cmake -DTT_TRAIN_INHERIT_COMPILER=OFF -DCMAKE_C_COMPILER=gcc-12 ...
+#   - If no CMakeCache.txt is found, falls back to CMake default detection.
+#   - CHECK_COMPILERS() validates the result (Clang 17+ or GCC 12+).
+#
+# When building as a tt-metal subproject (add_subdirectory), the parent
+# project's compiler is used and inheritance is skipped.
+#
+# Restored after deletion in a7618b9282 ("TT-Train: bump clang version
+# from 17 to 20 #36568").
+#
+# See: https://github.com/tenstorrent/tt-metal/issues/36993
+
+# Read the compiler used by tt-metal from its CMakeCache.txt and set it
+# as CMAKE_C_COMPILER/CMAKE_CXX_COMPILER so tt-train uses the same one.
+function(INHERIT_COMPILER_FROM_TT_METAL)
+    # Determine TT_METAL_HOME
+    if(DEFINED ENV{TT_METAL_HOME})
+        set(_tt_metal_home "$ENV{TT_METAL_HOME}")
+    else()
+        # Infer from directory structure: tt-metal/tt-train/ -> tt-metal/
+        set(_tt_metal_home "${CMAKE_CURRENT_SOURCE_DIR}/..")
+        if(NOT EXISTS "${_tt_metal_home}/tt_metal/CMakeLists.txt")
+            message(STATUS "Cannot infer tt-metal location; using CMake default compiler")
+            return()
+        endif()
+    endif()
+
+    # Search for CMakeCache.txt in known build directories
+    set(_cache_file "")
+    foreach(_build_dir "build" "build_Debug" "build_Release")
+        if(EXISTS "${_tt_metal_home}/${_build_dir}/CMakeCache.txt")
+            set(_cache_file "${_tt_metal_home}/${_build_dir}/CMakeCache.txt")
+            break()
+        endif()
+    endforeach()
+
+    if(NOT _cache_file)
+        message(STATUS "No tt-metal build found; using CMake default compiler")
+        return()
+    endif()
+
+    # Parse CMAKE_C_COMPILER and CMAKE_CXX_COMPILER from CMakeCache.txt
+    file(STRINGS "${_cache_file}" _c_compiler_line REGEX "^CMAKE_C_COMPILER:.*=")
+    file(STRINGS "${_cache_file}" _cxx_compiler_line REGEX "^CMAKE_CXX_COMPILER:.*=")
+
+    if(_c_compiler_line AND _cxx_compiler_line)
+        string(REGEX REPLACE "^CMAKE_C_COMPILER:[^=]*=(.*)" "\\1" _c_compiler "${_c_compiler_line}")
+        string(REGEX REPLACE "^CMAKE_CXX_COMPILER:[^=]*=(.*)" "\\1" _cxx_compiler "${_cxx_compiler_line}")
+
+        if(EXISTS "${_c_compiler}" AND EXISTS "${_cxx_compiler}")
+            message(STATUS "Inheriting compiler from tt-metal build (${_cache_file}):")
+            message(STATUS "  C compiler:   ${_c_compiler}")
+            message(STATUS "  C++ compiler: ${_cxx_compiler}")
+            set(CMAKE_C_COMPILER "${_c_compiler}" PARENT_SCOPE)
+            set(CMAKE_CXX_COMPILER "${_cxx_compiler}" PARENT_SCOPE)
+        else()
+            message(STATUS "Compiler paths from tt-metal CMakeCache not found on disk; using CMake default")
+        endif()
+    else()
+        message(STATUS "Could not parse compiler from ${_cache_file}; using CMake default")
+    endif()
+endfunction()
+
+function(CHECK_COMPILERS)
+    message(STATUS "Checking compilers: ${CMAKE_CXX_COMPILER_ID} ${CMAKE_CXX_COMPILER_VERSION}")
+
+    if(CMAKE_CXX_COMPILER_ID STREQUAL "Clang")
+        if(CMAKE_CXX_COMPILER_VERSION VERSION_LESS "17.0.0")
+            message(FATAL_ERROR "Clang-17 or higher is required")
+        endif()
+    elseif(CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
+        if(CMAKE_CXX_COMPILER_VERSION VERSION_LESS "12.0.0")
+            message(FATAL_ERROR "GCC-12 or higher is required")
+        endif()
+    else()
+        message(WARNING "Unsupported compiler: ${CMAKE_CXX_COMPILER_ID} ! Only Clang and GCC are supported")
+    endif()
+endfunction()

